pandas/core/series.py

```diff
@@ -4617,15 +4618,54 @@ def align(
             broadcast_axis=broadcast_axis,
         )
 
+    @overload
     def rename(
         self,
-        index=None,
+        index: Renamer | Hashable | None = ...,
```

Comment:
I think there may be a problem here that I uncovered when trying to use this in the MS stubs.  Here's a simple example:
```python
from typing import Mapping, Hashable, Any, Union


def myfun(m: Mapping[Hashable, Any]):
    pass


def myfun2(m: Union[Mapping[Hashable, Any], Hashable]):
    pass


myfun({1: 1})

myfun2({1: 1})
```
mypy 0.942 reports:
```
mapintint.py:14: error: Argument 1 to "myfun2" has incompatible type "Dict[int, int]"; expected "Union[Mapping[Hashable, Any], Hashable]"
```
So therefore if a user used the above, and wrote `    s5 = pd.Series([1, 2, 3]).rename({1: 10})
`, they'd see the mypy error.

I think mypy is wrong here.  If we change `Renamer` to use `Dict` instead of `Mapping`, then the mypy issue goes away.

Reply:
Or don't use `Union[Renamer, Hashable]` for `index` and have separate overloads for `Renamer` and `Hashable`

Reply:
> I think mypy is wrong here.

Seems like mypy wants to match dict with Hashable instead of matching it against Mapping:

> error: Argument 1 to "myfun2" has incompatible type "Dict[int, int]"; expected "Union[Mapping[Hashable, Any], Hashable]"
> note: Following member(s) of "Dict[int, int]" have conflicts:
> note:     __hash__: expected "Callable[[], int]", got "None"

I will look whether I find an existing mypy issue for it, if not i'll open one.

Reply:
Let's wait for mypy&pyright to figure out who is wrong before making changes :)

Reply:
[I'm surprised that Mapping doesn't require its keys to be Hashable.]

Reply:
The whole thing is pretty bizarre.  Consider this code:
```python
from typing import Hashable, Dict


def myfun(a: Dict[Hashable, str]):
    pass


myfun({1: "abc"})
myfun({"abc": "def"})
myfun({(1, 2): "xyz"})
d1 = {1: "abc"}
d2 = {"abc": "def"}
d3 = {(1, 2): "xyz"}
myfun(d1)
myfun(d2)
myfun(d3)


def myfun2(a: Hashable):
    pass


myfun2(1)
myfun2("abc")
myfun2((1, 2))
h1 = 1
h2 = "abc"
h3 = (1, 2)
myfun2(h1)
myfun2(h2)
myfun2(h3)
```
`mypy` complains about the second 3 calls to `myfun`, but not any of the calls to `myfun2`

Reply:
Do I understand the implications correctly: we should probably review all cases of invariance containers (Mapping and Sequence) and only use them if:

- the container is used without arguments or
- the container is used with Any.

Could also implement our own Dict/List-like protocol. Then we could allow Mapping-like objects that don't inherit from dict without loosening the type to Any.

Reply:
Or we know the exact specific types of the arguments in the container.  Using `Sequence` is fine - it is covariant.  Using `List[str]`, `Mapping[Hashable, Any]`, `Dict[Union[str,int], str]` is not.  So we can use `Mapping[str, str]`, or `Mapping` or `Mapping[Any, Any]`, etc.

Maybe open a separate PR for that - get it fixed there, then we accept this one.

Reply:
Note for future-self:

- Sub-classes of keys are not accepted by Mapping and Dict
- Sub-classes of values are not accepted by Dict but are accepted by Mapping
- List does not allow sub-classes of its items, Sequence does allow sub-classes of its items

Reply:
I used this in the MS stubs for now that I'm working on (not pushed or merged yet):
```python
Renamer = Union[Mapping[Any, Hashable], Callable[[Hashable], Hashable]]
```
I think this makes sense.  A `Mapping` has a key that must be hashable anyway, so making it map from `Any` to `Hashable` is probably what we want.

Reply:
Also, pyright now reports errors on this, so we may need a quick PR:  https://github.com/microsoft/pyright/issues/3258#issuecomment-1079969150

Reply:
I fear we have to go that way. I'll look into whether we can have a custom Mapping which has a non-Any upper bound but that might not be possible https://github.com/python/typing/pull/273

> A Mapping has a key that must be hashable anyway,

I think that according to typeshed that is not the case
https://github.com/python/typeshed/blob/c41034c3540c5abda10008dd6c17d8e6ff6cc634/stdlib/typing.pyi#L923
with `_KT = TypeVar("_KT")` doesn't enforce that it is hashable.

Reply:
I think I found a nicer solution that works with the latest pyright+mypy and is sanctions by [one of the typing gods](https://github.com/microsoft/pyright/issues/2344#issuecomment-927341908):

```py
# pyright: strict
from typing import Hashable, Mapping, TypeVar

T = TypeVar("T", bound=Hashable)

def test(x: Mapping[T, str]) -> None:
    ...

x1: Mapping[Hashable, str] = {}
x2: Mapping[str, str] = {}

test(x1)
test(x2)
```

Reply:
Interesting.  I tried that in the more complex MS stubs and it didn't work. 

We have to allow renames to things that aren't strings, and make sure that things like this work (Note that I changed the definition of `test`):

```python
from typing import Hashable, Mapping, TypeVar

T = TypeVar("T", bound=Hashable)


def test(x: Mapping[T, Hashable]) -> None:
    ...


x1: Mapping[Hashable, str] = {}
x2: Mapping[str, str] = {}
x3 = {1: 3}
x4 = {1: "a"}
x5 = {"a": "b"}
x6 = {"b": 3}
x7 = {"a": 1, 3: "b"}
test(x1)
test(x2)
test(x3)
test(x4)
test(x5)
test(x6)
test(x7)
```

While this works for `pyright` and `mypy` here, I could not get it working with the definitions of `rename` using arguments like `x1` through `x7` for `columns`.

In other words, this test was failing for `pyright` and `mypy`:
```python
def test_types_rename() -> None:
    df = pd.DataFrame(columns=["a"])
    col_map = {"a": "b"}
    df.rename(columns=col_map)
    df.rename(columns={"a": "b"})
    df.rename(columns={1: "b"})
    # Apparently all of these calls are accepted by pandas
    df.rename(columns={None: "b"})
    df.rename(columns={type("AnyObject")(): "b"})
    df.rename(columns={(2, 1): "b"})
```

Reply:
I'm sure there are some tricky cases to be covered with `rename` but when converting&appending your `test_types_rename` to match the example function `test`, I do not get any mypy&pyright errors:

```py
# pyright: strict
from typing import Hashable, Mapping, TypeVar

T = TypeVar("T", bound=Hashable)

def test(x: Mapping[T, Hashable]) -> None:
    ...

x1: Mapping[Hashable, str] = {}
x2: Mapping[str, str] = {}
x3 = {1: 3}
x4 = {1: "a"}
x5 = {"a": "b"}
x6 = {"b": 3}
x7 = {"a": 1, 3: "b"}
test(x1)
test(x2)
test(x3)
test(x4)
test(x5)
test(x6)
test(x7)

col_map = {"a": "b"}
test(col_map)
test({"a": "b"})
test({1: "b"})
test({None: "b"})  # None is hashable
test({object(): "b"})  # and I guess even object is hashable
test({(2, 1): "b"})
```

Reply:
Yes, I saw that, but it is the context of `rename()` that I couldn't get to work.  Might be some other issue with the version of the MS stubs I am working on.